Add 'stable' to the list of infinity version names.

mwkrentel opened the pull request:
Rename libunwind 1.5-head to 1.5-stable.

----------

I grep'd packages to find which other packages use version name
'stable' and found these with their maintainers.

hipsycl   [@nazavode]
hpx       [@msimberg, @albestro, @teonnik]
legion    [@pmccormick, @streichler]
neovim    [@albestro]
votca-*   [@junghans]

Anyone have a problem with this change?

Note that this will move the version higher in the order, probably
to where you really want it.

I also checked that no other package asks for libunwind 1.5-head.

Also, gasnet [@PHHargrove, @bonachea] uses version name 'main' for
branch stable, probably for the same reason that used 'head' for
libunwind, to get the order right.  Would you like to change that to
'stable'?

mwkrentel committed:
Add 'stable' to the list of infinity version names.
Rename libunwind 1.5-head to 1.5-stable.

bonachea commented:
Hi @mwkrentel -

> Also, gasnet [@PHHargrove, @bonachea] uses version name 'main' for branch stable, probably for the same reason that used 'head' for libunwind, to get the order right. Would you like to change that to 'stable'?

GASNet's `stable` branch falls version-wise between `develop` (changes most frequently) and `master` (changes least frequently). I deliberately mapped our git `stable` branch to Spack package version `main` which occupies the same position in the [sort order of infinity version names](https://spack.readthedocs.io/en/latest/packaging_guide.html). As package maintainer, IMO using a packaging version that matches the branch version ordering property is more important than matching the exact name of the underlying branch; because the discrepancy produces less surprising failures.

However if you'd consider inserting `stable` earlier in the list (I'd suggest between develop and main) _then_ I'd accept your proposed change to the GASNet package. Otherwise I'd prefer it left as-is.

msimberg commented:
I think I'm ok with this change, though I wasn't familiar with "infinity versions" from before. It just means that they compare as newer versions than any other ordinary version? If yes, this seems reasonable.

mwkrentel commented:
@bonachea To me, 'stable' suggests the most conservative, least
changing version, so I would prefer to keep it at the bottom.

But I agree with you that order is more important, so we'll keep
gasnet at 'main'.  I'm certainly not going to push it if you don't want it.

mwkrentel commented:
> It just means that they compare as newer versions than any other ordinary version? 

@msimberg Yes, that's what it means, spack calls these "infinity versions."

For the case of libunwind, 1.5.0 is a fixed, released version.
1.5-stable is the branch that starts at 1.5.0 with conservative bug
fixes, so it comes after 1.5.0.  libunwind calls it 'stable', so ....

Hmm, I missed this on my first pass.  This change will move stable
ahead of 1.7.1, and below master.  That means that some of your
version checks:

```
with when('@1.7:'):
    depends_on('cmake@3.18.0:', type='build')
    depends_on('boost@1.71.0:')
    depends_on('asio@1.12.0:')
    conflicts('%gcc@:6')
    conflicts('%clang@:6')
```

will now apply to stable.  Is that going to be a problem?  That is, is
stable closer to <= 1.1.0 or >= 1.7.1 in terms of your restrictions?
It's easy to change the range if needed.

My apologies, I missed that.

msimberg commented:
Thanks for the explanation @mwkrentel!

> will now apply to stable. Is that going to be a problem? That is, is stable closer to <= 1.1.0 or >= 1.7.1 in terms of your restrictions? It's easy to change the range if needed.

It's the latter, i.e. it's a sort of snapshot of master. Not quite the same as master, which is a development branch, but usually newer than the latest release. So it sounds like an improvement!

mwkrentel commented:
@msimberg Great!  That's what I was thinking based on the order that the
versions are listed (master, stable, 1.7.1).  Btw, the order of
calling version() does not determine the real order.  Spack sorts them
its own way regardless of how they are called.

Btw, after this patch, if the constraints for master and stable are
identical, you can combine them with `when('@master,stable')`
or `when('@stable:')`.  But that's just a style thing.

mwkrentel commented:
Closed/reopened PR to restart the tests.
The CI failed and I have no idea why or how to fix.
Anyone ??

Well, that didn't help.   CI still fails and I don't know why.  :-(

streichler commented:
Chiming in late, but this ordering matches Legion's use of "stable", so no complaints here.

scottwittenburg commented:
@mwkrentel Can you rebase this PR on the latest spack `develop`?  I'm not sure what commit fixed the broken `ascent` package, but PRs based on a more recent `develop` don't seem to be suffering from the problem this PR encountered in gitlab CI.

mwkrentel committed:
Add stable to the infinite version list in packaging_guide.rst.

mwkrentel commented:
@scottwittenburg I rebased with develop.  It was only 40 commits and
no conflicts, but I now get the magic green check.

I also added stable to the list in packaging_guide.rst in the docs (I previously forgot).
Maybe it was just karma working its magic.  :-)

So now I'm looking for a review and approval and merge.

Thanks!

albestro commented:
Sorry for coming late to the party.

Here I'm (trying) to report about `neovim`, since on `hpx` @msimberg is way more authoritative than me.

I've just "adopted" `neovim` package in spack, so I don't really know exactly the dynamics. 
Anyway, by quickly checking the repo:
- looking at [how CI works](https://github.com/jamessan/neovim/blob/9055ec5792d29eb8ad5e5c20708ffcb980ad8609/.github/workflows/release.yml#L176-L183), it seems `stable` tag is pushed as soon as a new tag release, e.g. `0.6.1`, is pushed.
- it seems `neovim` uses `stable` tag for `snap`, e.g. according to https://github.com/neovim/neovim/issues/14464, and by looking at its [doc](https://snapcraft.io/docs/channels#heading--risk-levels), it sounds consistent with the previous point.

For these reasons, it should be the right choice also for `neovim`. Thanks!
